File: Front-end/mentor_management_system/src/app/services/leave-permission.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class LeavePermissionService {

  private apiUrl = 'http://localhost:8081/leave';

  constructor(private http: HttpClient) { }

  // Apply for leave (Student)
  applyLeave(leaveData: any): Observable<any> {
    return this.http.post(`${this.apiUrl}/apply`, leaveData);
  }

  // Get all leaves for a specific student (Student)
  getStudentLeaves(studentId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/student/${studentId}`);
  }

  // Get all leaves (Admin)
  getAllLeaves(): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/all`);
  }

  // Get leaves for a specific mentor's students
  getLeavesByMentorId(mentorId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/mentor/${mentorId}`);
  }

  // Get single leave by ID
  getLeaveById(id: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/${id}`);
  }

  // Approve leave (Admin/Mentor)
  approveLeave(id: number): Observable<any> {
    return this.http.put(`${this.apiUrl}/approve/${id}`, {});
  }

  // Reject leave (Admin/Mentor)
  rejectLeave(id: number): Observable<any> {
    return this.http.put(`${this.apiUrl}/reject/${id}`, {});
  }
}

File: Backend/Mentor_Management_System/src/main/java/com/example/demo/service/LeavePermissionService.java
package com.example.demo.service;

import java.util.List;
import com.example.demo.model.LeavePermission;

public interface LeavePermissionService {

    LeavePermission applyLeave(LeavePermission leave);

    List<LeavePermission> getAllLeaves();

    LeavePermission approveLeave(Long id);

    LeavePermission rejectLeave(Long id);

    List<LeavePermission> getLeavesByStudentId(Long studentId);

    List<LeavePermission> getLeavesByMentorId(Integer mentorId);

    LeavePermission getLeaveById(Long id);
}